btree/btree.c

```diff
@@ -3504,67 +3569,145 @@ static void fkvv_done(struct slot *slot, struct m0_be_tx *tx, bool modified)
 
-	if (idx == 0){
-		total_val_size = (*fkvv_val_offset_get(slot->s_node, h->fkvv_used-1));
-	} else {
-		total_val_size = (*fkvv_val_offset_get(slot->s_node, h->fkvv_used-1)) - (*fkvv_val_offset_get(slot->s_node, idx -1));
+	h->fkvv_used++;
+}
+
+static void fkvv_make_leaf(struct slot *slot, struct m0_be_tx *tx)
+{
+	struct fkvv_head *h = fkvv_data(slot->s_node);
+	uint32_t         *curr_val_offset;
+	uint32_t         *prev_val_offset;
+	uint32_t         *last_val_offset;
+	uint32_t          new_val_offset;
+	void             *key_addr;
+	void             *val_addr;
+	int               unit_key_offset_rsize;
+	int               total_key_size;
+	int               total_val_size;
+	int               new_val_size;
+	int               idx       = slot->s_idx;
+	int               last_idx  = h->fkvv_used - 1;
+	int               i;
+
+	new_val_size =  m0_vec_count(&slot->s_rec.r_val.ov_vec);
+
+	if (slot->s_idx == 0)
+		new_val_offset  = new_val_size;
+	else {
+		prev_val_offset = fkvv_val_offset_get(slot->s_node, idx - 1);
+		new_val_offset  = *prev_val_offset + new_val_size;
+	}
+
+	if (slot->s_idx == h->fkvv_used) {
+		h->fkvv_used++;
+		curr_val_offset  = fkvv_val_offset_get(slot->s_node, idx);
+		*curr_val_offset = new_val_offset;
+		return;
 	}
 
-	memmove(addr_key_at_idx + key_offset_area_size, addr_key_at_idx, total_key_size);
-	memmove(addr_val_at_last_idx - incoming_val_size, addr_val_at_last_idx , total_val_size);
+	last_val_offset        = fkvv_val_offset_get(slot->s_node, last_idx);
+	unit_key_offset_rsize  = h->fkvv_ksize + OFFSET_SIZE;
+
+	key_addr               = fkvv_key(slot->s_node, idx);
+	val_addr               = fkvv_val(slot->s_node, last_idx);
+
+	total_key_size         = (unit_key_offset_rsize) * (h->fkvv_used - idx);
+	total_val_size         = (idx == 0) ? *last_val_offset :
+				 *last_val_offset - *prev_val_offset;
+
+	memmove(key_addr + unit_key_offset_rsize, key_addr, total_key_size);
+	memmove(val_addr - new_val_size, val_addr , total_val_size);
 
 	h->fkvv_used++;
 
-	/* update value offset */
-	fkvv_val_offset_set(slot->s_node, idx, offset_for_new_val);
-	for(i = idx+1; i < h->fkvv_used; i++)
-	{
-		val_offset_at_i = fkvv_val_offset_get(slot->s_node, i);
-		*val_offset_at_i = *val_offset_at_i + incoming_val_size;
+	/* Update offeset values */
+	curr_val_offset  = fkvv_val_offset_get(slot->s_node, idx);
+	*curr_val_offset = new_val_offset;
+
+	for (i = idx + 1; i < h->fkvv_used; i++) {
+		curr_val_offset  = fkvv_val_offset_get(slot->s_node, i);
+		*curr_val_offset = *curr_val_offset + new_val_size;
 	}
-	#endif
 }
 
-static bool fkvv_find(struct slot *slot, const struct m0_btree_key *key)
+static void fkvv_make(struct slot *slot, struct m0_be_tx *tx)
 {
-	return 0;
+	(IS_INTERNAL_NODE(slot->s_node)) ? fkvv_make_internal(slot, tx)
+					 : fkvv_make_leaf(slot, tx);
+}
+
+static bool fkvv_find(struct slot *slot, const struct m0_btree_key *find_key)
```

Comment:
for fkvv_find() and ff_find() the code is common as key size is constant but in case of variable key size I thing code may change little bit  still if we want to have one common function node_find we can do it but need to check with Yash.

Reply:
We can have a common function for this, but certain things will needed to be changed.
Like we cannot use h->ff_ksize for sizes, we need to use function.
Pre-condition about size needs to rethought of.

Reply:
I have not made this change as if we make node_find as common function, in case of variable key , we need to call function for getting key size at specific index but such functions will not be present for fixed key size node format, I guess It is better to have  separate implementation based on node_format.

Reply:
I am not sure why are we not able to do this? Having a common code will help us - this is just like the generic_move() code which is supposed to be used in all the node formats. We can talk about this if needed.

Reply:
Yes right , I have made this changes Please check, Earlier I thought we need to define some other routine for all node_type to get key size at specific index but we can make use of node_key.